

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Variables and arithmetic operators &#8212; Python For Biochemists</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'chapters/ExamplesSolutionsChapters';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../WelcomeToPythonForBiochemists.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Python For Biochemists - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="Python For Biochemists - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../WelcomeToPythonForBiochemists.html">
                    Welcome to Python for Biochemists
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="JupyterNotebooksGettingStarted.html">Jupyter Notebooks: Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="VariablesArithmeticOperators.html">Variables and Arithmetic Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="DataTypesIndicesSlices.html">Data Types, Index, and Slice</a></li>
<li class="toctree-l1"><a class="reference internal" href="ForLoopsLogicStatementsFunctions.html">For Loops, Logic Statements, and Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="Libraries.html">Libraries</a></li>
<li class="toctree-l1"><a class="reference internal" href="DrawingGraphs.html">Drawing Graphs</a></li>
<li class="toctree-l1"><a class="reference internal" href="PandasForTabularData.html">Pandas For Working With Tabular Data</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fchapters/ExamplesSolutionsChapters.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/chapters/ExamplesSolutionsChapters.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1><span style="color:blue">Variables and arithmetic operators</span></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#"><span style="color:blue">Variables and arithmetic operators</span></a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-types-index-and-slice"><span style="color:blue">Data types, index, and slice</span></a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#for-loops-logic-statements-and-functions"><span style="color:blue">For loops, logic statements, and functions</span></a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#drawing-graphs"><span style="color:blue">Drawing graphs</span></a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#pandas-for-working-with-tabular-data"><span style="color:blue">Pandas For Working With Tabular Data</span></a></li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="variables-and-arithmetic-operators">
<h1><span style="color:blue">Variables and arithmetic operators</span><a class="headerlink" href="#variables-and-arithmetic-operators" title="Permalink to this heading">#</a></h1>
<hr class="docutils" />
<p><strong>How many grams of solid NaOH (40.0 g/mol) are required to prepare 500 ml of a 40 mM solution?</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">V</span> <span class="o">=</span> <span class="mf">0.5</span>   <span class="c1">#l</span>
<span class="n">M</span> <span class="o">=</span> <span class="mf">0.04</span>   <span class="c1">#mol/l</span>
<span class="n">MW</span> <span class="o">=</span> <span class="mf">40.0</span>   <span class="c1">#g/mol</span>

<span class="n">m</span> <span class="o">=</span> <span class="p">(</span><span class="n">V</span> <span class="o">*</span> <span class="n">M</span><span class="p">)</span> <span class="o">*</span> <span class="n">MW</span>   <span class="c1">#(l * mol/l) * g/mol = g</span>

<span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>   <span class="c1">#print the value that we calculated</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.8
</pre></div>
</div>
</div>
</div>
<p><strong>How many ml of a 1 M Tris solution do you need to make 5 ml of a 0.25 M solution?</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Vfinal</span> <span class="o">=</span> <span class="mi">5</span>   <span class="c1">#ml</span>
<span class="n">Cfinal</span> <span class="o">=</span> <span class="mf">0.25</span>   <span class="c1">#mol/l</span>
<span class="n">Cinitial</span> <span class="o">=</span> <span class="mi">1</span>   <span class="c1">#mol/l</span>

<span class="n">Vinitial</span> <span class="o">=</span> <span class="p">(</span><span class="n">Cfinal</span> <span class="o">*</span> <span class="n">Vfinal</span><span class="p">)</span> <span class="o">/</span> <span class="n">Cinitial</span>   <span class="c1">#(mol/l * ml) / mol/l = ml</span>

<span class="nb">print</span><span class="p">(</span><span class="n">Vinitial</span><span class="p">)</span>   <span class="c1">#print the value that we calculated</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1.25
</pre></div>
</div>
</div>
</div>
<p><strong>You need 250 ml 50 mM NaCl. You have a 2 M NaCl stock solution. How many ml of 2 M NaCl do you need?</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Vfinal</span> <span class="o">=</span> <span class="mf">0.25</span>   <span class="c1">#l</span>
<span class="n">Cfinal</span> <span class="o">=</span> <span class="mf">0.05</span>   <span class="c1">#mol/l</span>
<span class="n">Cinitial</span> <span class="o">=</span> <span class="mi">2</span>   <span class="c1">#mol/l</span>

<span class="n">Vinitial</span> <span class="o">=</span> <span class="p">(</span><span class="n">Cfinal</span> <span class="o">*</span> <span class="n">Vfinal</span><span class="p">)</span> <span class="o">/</span> <span class="n">Cinitial</span>   <span class="c1">#(l * mol/l) / mol/l = l</span>
<span class="n">Vinitialml</span> <span class="o">=</span> <span class="n">Vinitial</span><span class="o">*</span><span class="mi">1000</span>   <span class="c1">#l/1000 = ml</span>

<span class="nb">print</span><span class="p">(</span><span class="n">Vinitialml</span><span class="p">)</span>   <span class="c1">#print the value that we calculated</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>6.25
</pre></div>
</div>
</div>
</div>
<p><strong>You have 1 g of ATP (551.1 g/mol). You need a 100 mM solution. How much solution (in ml) can you make?</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">m</span> <span class="o">=</span> <span class="mi">1</span>   <span class="c1">#g</span>
<span class="n">M</span> <span class="o">=</span> <span class="mf">0.1</span>   <span class="c1">#mol/l</span>
<span class="n">MW</span> <span class="o">=</span> <span class="mf">551.1</span>   <span class="c1">#g/mol</span>

<span class="n">V</span> <span class="o">=</span> <span class="p">(</span><span class="n">m</span> <span class="o">/</span> <span class="n">MW</span><span class="p">)</span> <span class="o">/</span> <span class="n">M</span>   <span class="c1">#(g * g/mol) * mol/l = l</span>
<span class="n">Vml</span> <span class="o">=</span> <span class="n">V</span><span class="o">*</span><span class="mi">1000</span>   <span class="c1">#l/1000 = ml</span>

<span class="nb">print</span><span class="p">(</span><span class="n">Vml</span><span class="p">)</span>   <span class="c1">#print the value that we calculated</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>18.145527127563053
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="data-types-index-and-slice">
<h1><span style="color:blue">Data types, index, and slice</span><a class="headerlink" href="#data-types-index-and-slice" title="Permalink to this heading">#</a></h1>
<hr class="docutils" />
<p><strong>Determine the data type, length, and number of tryptophan residues for this LRRK2 protein sequence containing one letter code amino acids. Select residue 2019. The G2019S mutation in LRRK2 is the most common genetic determinant of Parkinson’s disease identified to date. It lies in the protein’s kinase domain. Select this kinase domain, it includes residue 1879 to residue 2138.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">protseqLRRK2</span> <span class="o">=</span> <span class="s2">&quot;MASGSCQGCEEDEETLKKLIVRLNNVQEGKQIETLVQILEDLLVFTYSERASKLFQGKNIHVPLLIVLDSYMRVASVQQVGWSLLCKLIEVCPGTMQSLMGPQDVGNDWEVLGVHQLILKMLTVHNASVNLSVIGLKTLDLLLTSGKITLLILDEESDIFMLIFDAMHSFPANDEVQKLGCKALHVLFERVSEEQLTEFVENKDYMILLSALTNFKDEEEIVLHVLHCLHSLAIPCNNVEVLMSGNVRCYNIVVEAMKAFPMSERIQEVSCCLLHRLTLGNFFNILVLNEVHEFVVKAVQQYPENAALQISALSCLALLTETIFLNQDLEEKNENQENDDEGEEDKLFWLEACYKALTWHRKNKHVQEAACWALNNLLMYQNSLHEKIGDEDGHFPAHREVMLSMLMHSSSKEVFQASANALSTLLEQNVNFRKILLSKGIHLNVLELMQKHIHSPEVAESGCKMLNHLFEGSNTSLDIMAAVVPKILTVMKRHETSLPVQLEALRAILHFIVPGMPEESREDTEFHHKLNMVKKQCFKNDIHKLVLAALNRFIGNPGIQKCGLKVISSIVHFPDALEMLSLEGAMDSVLHTLQMYPDDQEIQCLGLSLIGYLITKKNVFIGTGHLLAKILVSSLYRFKDVAEIQTKGFQTILAILKLSASFSKLLVHHSFDLVIFHQMSSNIMEQKDQQFLNLCCKCFAKVAMDDYLKNVMLERACDQNNSIMVECLLLLGADANQAKEGSSLICQVCEKESSPKLVELLLNSGSREQDVRKALTISIGKGDSQIISLLLRRLALDVANNSICLGGFCIGKVEPSWLGPLFPDKTSNLRKQTNIASTLARMVIRYQMKSAVEEGTASGSDGNFSEDVLSKFDEWTFIPDSSMDSVFAQSDDLDSEGSEGSFLVKKKSNSISVGEFYRDAVLQRCSPNLQRHSNSLGPIFDHEDLLKRKRKILSSDDSLRSSKLQSHMRHSDSISSLASEREYITSLDLSANELRDIDALSQKCCISVHLEHLEKLELHQNALTSFPQQLCETLKSLTHLDLHSNKFTSFPSYLLKMSCIANLDVSRNDIGPSVVLDPTVKCPTLKQFNLSYNQLSFVPENLTDVVEKLEQLILEGNKISGICSPLRLKELKILNLSKNHISSLSENFLEACPKVESFSARMNFLAAMPFLPPSMTILKLSQNKFSCIPEAILNLPHLRSLDMSSNDIQYLPGPAHWKSLNLRELLFSHNQISILDLSEKAYLWSRVEKLHLSHNKLKEIPPEIGCLENLTSLDVSYNLELRSFPNEMGKLSKIWDLPLDELHLNFDFKHIGCKAKDIIRFLQQRLKKAVPYNRMKLMIVGNTGSGKTTLLQQLMKTKKSDLGMQSATVGIDVKDWPIQIRDKRKRDLVLNVWDFAGREEFYSTHPHFMTQRALYLAVYDLSKGQAEVDAMKPWLFNIKARASSSPVILVGTHLDVSDEKQRKACMSKITKELLNKRGFPAIRDYHFVNATEESDALAKLRKTIINESLNFKIRDQLVVGQLIPDCYVELEKIILSERKNVPIEFPVIDRKRLLQLVRENQLQLDENELPHAVHFLNESGVLLHFQDPALQLSDLYFVEPKWLCKIMAQILTVKVEGCPKHPKGIISRRDVEKFLSKKRKFPKNYMSQYFKLLEKFQIALPIGEEYLLVPSSLSDHRPVIELPHCENSEIIIRLYEMPYFPMGFWSRLINRLLEISPYMLSGRERALRPNRMYWRQGIYLNWSPEAYCLVGSEVLDNHPESFLKITVPSCRKGCILLGQVVDHIDSLMEEWFPGLLEIDICGEGETLLKKWALYSFNDGEEHQKILLDDLMKKAEEGDLLVNPDQPRLTIPISQIAPDLILADLPRNIMLNNDELEFEQAPEFLLGDGSFGSVYRAAYEGEEVAVKIFNKHTSLRLLRQELVVLCHLHHPSLISLLAAGIRPRMLVMELASKGSLDRLLQQDKASLTRTLQHRIALHVADGLRYLHSAMIIYRDLKPHNVLLFTLYPNAAIIAKIADYGIAQYCCRMGIKTSEGTPGFRAPEVARGNVIYNQQADVYSFGLLLYDILTTGGRIVEGLKFPNEFDELEIQGKLPDPVKEYGCAPWPMVEKLIKQCLKENPQERPTSAQVFDILNSAELVCLTRRILLPKNVIVECMVATHHNSRNASIWLGCGHTDRGQLSFLDLNTEGYTSEEVADSRILCLALVHLPVEKESWIVSGTQSGTLLVINTEDGKKRHTLEKMTDSVTCLYCNSFSKQSKQKNFLLVGTADGKLAIFEDKTVKLKGAAPLKILNIGNVSTPLMCLSESTNSTERNVMWGGCGTKIFSFSNDFTIQKLIETRTSQLFSYAAFSDSNIITVVVDTALYIAKQNSPVVEVWDKKTEKLCGLIDCVHFLREVMVKENKESKHKMSYSGRVKTLCLQKNTALWIGTGGGHILLLDLSTRRLIRVIYNFCNSVRVMMTAQLGSLKNVMLVLGYNRKNTEGTQKQKEIQSCLTVWDINLPHEVQNLEKHIEVRKELAEKMRRTSVE&quot;</span>   <span class="c1">#create a string using double quotation marks</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">type_protseqLRRK2</span> <span class="o">=</span> <span class="nb">type</span><span class="p">(</span><span class="n">protseqLRRK2</span><span class="p">)</span>   <span class="c1">#determine the data type</span>
<span class="n">len_protseqLRRK2</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">protseqLRRK2</span><span class="p">)</span>   <span class="c1">#determine the length of the string</span>
<span class="n">Wcount_protseqLRRK2</span> <span class="o">=</span> <span class="n">protseqLRRK2</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;W&quot;</span><span class="p">)</span>   <span class="c1">#count the number of times W appears in the string</span>
<span class="n">res2019_protseqLRRK2</span> <span class="o">=</span> <span class="n">protseqLRRK2</span><span class="p">[</span><span class="mi">2018</span><span class="p">]</span>   <span class="c1">#select residue 2019 (2018 as the first item has index 0)</span>
<span class="n">kinase_protseqLRRK2</span> <span class="o">=</span> <span class="n">protseqLRRK2</span><span class="p">[</span><span class="mi">1878</span><span class="p">:</span><span class="mi">2138</span><span class="p">]</span>   <span class="c1">#select residues 1879 (1878 as the first item has index 0) to 2138 (the first item has index 0; however, it goes up to but does not include item 2138)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;This variable has data type&quot;</span><span class="p">,</span> <span class="n">type_protseqLRRK2</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span> 
      <span class="s2">&quot;The length of the sequence is&quot;</span><span class="p">,</span> <span class="n">len_protseqLRRK2</span><span class="p">,</span> <span class="s2">&quot;residues .&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Trp appears&quot;</span><span class="p">,</span> <span class="n">Wcount_protseqLRRK2</span><span class="p">,</span> <span class="s2">&quot;times .&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Amino acid 2019 is a&quot;</span><span class="p">,</span> <span class="n">res2019_protseqLRRK2</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span>
      <span class="s2">&quot;The sequence of the LRRK2 kinase domain is&quot;</span><span class="p">,</span> <span class="n">kinase_protseqLRRK2</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>This variable has data type &lt;class &#39;str&#39;&gt; . The length of the sequence is 2527 residues . Trp appears 26 times . Amino acid 2019 is a G . The sequence of the LRRK2 kinase domain is QAPEFLLGDGSFGSVYRAAYEGEEVAVKIFNKHTSLRLLRQELVVLCHLHHPSLISLLAAGIRPRMLVMELASKGSLDRLLQQDKASLTRTLQHRIALHVADGLRYLHSAMIIYRDLKPHNVLLFTLYPNAAIIAKIADYGIAQYCCRMGIKTSEGTPGFRAPEVARGNVIYNQQADVYSFGLLLYDILTTGGRIVEGLKFPNEFDELEIQGKLPDPVKEYGCAPWPMVEKLIKQCLKENPQERPTSAQVFDILNSAELV .
</pre></div>
</div>
</div>
</div>
<p><strong>Determine the number of “key : value” pairs in the following dictionary with genetic codes, with key = code and value = amino acid. Retrieve the genetic code translation for TTC.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">GeneticCode</span> <span class="o">=</span> <span class="p">{</span> 
        <span class="s1">&#39;ATA&#39;</span><span class="p">:</span><span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;ATC&#39;</span><span class="p">:</span><span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;ATT&#39;</span><span class="p">:</span><span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;ATG&#39;</span><span class="p">:</span><span class="s1">&#39;M&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;ACA&#39;</span><span class="p">:</span><span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;ACC&#39;</span><span class="p">:</span><span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;ACG&#39;</span><span class="p">:</span><span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;ACT&#39;</span><span class="p">:</span><span class="s1">&#39;T&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;AAC&#39;</span><span class="p">:</span><span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;AAT&#39;</span><span class="p">:</span><span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;AAA&#39;</span><span class="p">:</span><span class="s1">&#39;K&#39;</span><span class="p">,</span> <span class="s1">&#39;AAG&#39;</span><span class="p">:</span><span class="s1">&#39;K&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;AGC&#39;</span><span class="p">:</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;AGT&#39;</span><span class="p">:</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;AGA&#39;</span><span class="p">:</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;AGG&#39;</span><span class="p">:</span><span class="s1">&#39;R&#39;</span><span class="p">,</span>                  
        <span class="s1">&#39;CTA&#39;</span><span class="p">:</span><span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;CTC&#39;</span><span class="p">:</span><span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;CTG&#39;</span><span class="p">:</span><span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;CTT&#39;</span><span class="p">:</span><span class="s1">&#39;L&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;CCA&#39;</span><span class="p">:</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;CCC&#39;</span><span class="p">:</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;CCG&#39;</span><span class="p">:</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;CCT&#39;</span><span class="p">:</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;CAC&#39;</span><span class="p">:</span><span class="s1">&#39;H&#39;</span><span class="p">,</span> <span class="s1">&#39;CAT&#39;</span><span class="p">:</span><span class="s1">&#39;H&#39;</span><span class="p">,</span> <span class="s1">&#39;CAA&#39;</span><span class="p">:</span><span class="s1">&#39;Q&#39;</span><span class="p">,</span> <span class="s1">&#39;CAG&#39;</span><span class="p">:</span><span class="s1">&#39;Q&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;CGA&#39;</span><span class="p">:</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;CGC&#39;</span><span class="p">:</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;CGG&#39;</span><span class="p">:</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;CGT&#39;</span><span class="p">:</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;GTA&#39;</span><span class="p">:</span><span class="s1">&#39;V&#39;</span><span class="p">,</span> <span class="s1">&#39;GTC&#39;</span><span class="p">:</span><span class="s1">&#39;V&#39;</span><span class="p">,</span> <span class="s1">&#39;GTG&#39;</span><span class="p">:</span><span class="s1">&#39;V&#39;</span><span class="p">,</span> <span class="s1">&#39;GTT&#39;</span><span class="p">:</span><span class="s1">&#39;V&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;GCA&#39;</span><span class="p">:</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;GCC&#39;</span><span class="p">:</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;GCG&#39;</span><span class="p">:</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;GCT&#39;</span><span class="p">:</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;GAC&#39;</span><span class="p">:</span><span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;GAT&#39;</span><span class="p">:</span><span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;GAA&#39;</span><span class="p">:</span><span class="s1">&#39;E&#39;</span><span class="p">,</span> <span class="s1">&#39;GAG&#39;</span><span class="p">:</span><span class="s1">&#39;E&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;GGA&#39;</span><span class="p">:</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;GGC&#39;</span><span class="p">:</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">:</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;GGT&#39;</span><span class="p">:</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;TCA&#39;</span><span class="p">:</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;TCC&#39;</span><span class="p">:</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;TCG&#39;</span><span class="p">:</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;TCT&#39;</span><span class="p">:</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;TTC&#39;</span><span class="p">:</span><span class="s1">&#39;F&#39;</span><span class="p">,</span> <span class="s1">&#39;TTT&#39;</span><span class="p">:</span><span class="s1">&#39;F&#39;</span><span class="p">,</span> <span class="s1">&#39;TTA&#39;</span><span class="p">:</span><span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;TTG&#39;</span><span class="p">:</span><span class="s1">&#39;L&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;TAC&#39;</span><span class="p">:</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;TAT&#39;</span><span class="p">:</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;TAA&#39;</span><span class="p">:</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39;TAG&#39;</span><span class="p">:</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> 
        <span class="s1">&#39;TGC&#39;</span><span class="p">:</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;TGT&#39;</span><span class="p">:</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;TGA&#39;</span><span class="p">:</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39;TGG&#39;</span><span class="p">:</span><span class="s1">&#39;W&#39;</span><span class="p">,</span> 
    <span class="p">}</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">len_GeneticCode</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">GeneticCode</span><span class="p">)</span>   <span class="c1">#determine the length of the dictionary</span>

<span class="n">GeneticCode_TTC</span> <span class="o">=</span> <span class="n">GeneticCode</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;TTC&#39;</span><span class="p">)</span>   <span class="c1">#look up the cleavage site stored for NotI</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;This dictionary contains&quot;</span><span class="p">,</span> <span class="n">len_GeneticCode</span><span class="p">,</span> <span class="s2">&quot;pairs .&quot;</span><span class="p">,</span>
      <span class="s2">&quot;The genetic code translation stored for TTC is&quot;</span><span class="p">,</span> <span class="n">GeneticCode_TTC</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>This dictionary contains 64 pairs . The genetic code translation stored for TTC is F .
</pre></div>
</div>
</div>
</div>
<p><strong>Determine the data type and length for this list with amino acids. Sort the list in alphabetical order.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">AA3Letter</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ALA&quot;</span><span class="p">,</span> <span class="s2">&quot;ARG&quot;</span><span class="p">,</span> <span class="s2">&quot;ASN&quot;</span><span class="p">,</span> <span class="s2">&quot;ASP&quot;</span><span class="p">,</span> <span class="s2">&quot;CYS&quot;</span><span class="p">,</span> <span class="s2">&quot;GLN&quot;</span><span class="p">,</span> <span class="s2">&quot;GLU&quot;</span><span class="p">,</span> <span class="s2">&quot;GLY&quot;</span><span class="p">,</span> <span class="s2">&quot;HIS&quot;</span><span class="p">,</span> <span class="s2">&quot;ILE&quot;</span><span class="p">,</span> <span class="s2">&quot;LEU&quot;</span><span class="p">,</span> <span class="s2">&quot;LYS&quot;</span><span class="p">,</span> <span class="s2">&quot;MET&quot;</span><span class="p">,</span> <span class="s2">&quot;PHE&quot;</span><span class="p">,</span> <span class="s2">&quot;PRO&quot;</span><span class="p">,</span> <span class="s2">&quot;SER&quot;</span><span class="p">,</span> <span class="s2">&quot;THR&quot;</span><span class="p">,</span> <span class="s2">&quot;TRP&quot;</span><span class="p">,</span> <span class="s2">&quot;TYR&quot;</span><span class="p">,</span> <span class="s2">&quot;VAL&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">type_AA3Letter</span> <span class="o">=</span> <span class="nb">type</span><span class="p">(</span><span class="n">AA3Letter</span><span class="p">)</span>   <span class="c1">#determine and print the data type</span>

<span class="n">len_AA3Letter</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">AA3Letter</span><span class="p">)</span>   <span class="c1">#determine the length of the list</span>
<span class="n">AA3Lettersorted</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">AA3Letter</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="nb">str</span><span class="o">.</span><span class="n">lower</span><span class="p">)</span>   <span class="c1">#sort the list in ascending order</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;This variable has data type&quot;</span><span class="p">,</span> <span class="n">type_AA3Letter</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span> 
      <span class="s2">&quot;This list contains&quot;</span><span class="p">,</span> <span class="n">len_AA3Letter</span><span class="p">,</span> <span class="s2">&quot;amino acids .&quot;</span><span class="p">,</span>
      <span class="s2">&quot;After sorting the concentrations in ascending order, this list becomes&quot;</span><span class="p">,</span> <span class="n">AA3Lettersorted</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>This variable has data type &lt;class &#39;list&#39;&gt; . This list contains 20 amino acids . After sorting the concentrations in ascending order, this list becomes [&#39;ALA&#39;, &#39;ARG&#39;, &#39;ASN&#39;, &#39;ASP&#39;, &#39;CYS&#39;, &#39;GLN&#39;, &#39;GLU&#39;, &#39;GLY&#39;, &#39;HIS&#39;, &#39;ILE&#39;, &#39;LEU&#39;, &#39;LYS&#39;, &#39;MET&#39;, &#39;PHE&#39;, &#39;PRO&#39;, &#39;SER&#39;, &#39;THR&#39;, &#39;TRP&#39;, &#39;TYR&#39;, &#39;VAL&#39;] .
</pre></div>
</div>
</div>
</div>
<p><strong>Define the EcoRI DNA recognition sequence (GAATTC) as a string. Determine length and number of adenine bases for this sequence.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DNAseqEcoRI</span> <span class="o">=</span> <span class="s2">&quot;GAATTC&quot;</span>   <span class="c1">#create a string using double quotation marks</span>

<span class="n">len_DNAseqEcoRI</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">DNAseqEcoRI</span><span class="p">)</span>   <span class="c1">#determine the length of the string</span>
<span class="n">Acount_DNAseqEcoRI</span> <span class="o">=</span> <span class="n">DNAseqEcoRI</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>   <span class="c1">#count the number of times A appears in the string</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The length of the sequence is&quot;</span><span class="p">,</span> <span class="n">len_DNAseqEcoRI</span><span class="p">,</span> <span class="s2">&quot;residues .&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Adenosine appears&quot;</span><span class="p">,</span> <span class="n">Acount_DNAseqEcoRI</span><span class="p">,</span> <span class="s2">&quot;times .&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The length of the sequence is 6 residues . Adenosine appears 2 times .
</pre></div>
</div>
</div>
</div>
<p><strong>Calculate the melting temperature of the primer with sequence “GACTGCGTTAGGATTGGC”. The melting temperature (°C) can be computed as 64.9 + 41 x (GC - 16.4)/N, with GC the total number of G and C bases in the primer and N the primer length.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">primer</span> <span class="o">=</span> <span class="s2">&quot;GACTGCGTTAGGATTGGC&quot;</span>   <span class="c1">#create a string using double quotation marks</span>

<span class="n">N_primer</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">primer</span><span class="p">)</span>   <span class="c1">#determine the length of the string</span>
<span class="n">G_primer</span> <span class="o">=</span> <span class="n">primer</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">)</span>   <span class="c1">#count the number of times G appears in the string</span>
<span class="n">C_primer</span> <span class="o">=</span> <span class="n">primer</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>   <span class="c1">#count the number of times C appears in the string</span>
<span class="n">GC_primer</span> <span class="o">=</span> <span class="n">G_primer</span> <span class="o">+</span> <span class="n">C_primer</span>   <span class="c1">#calculate the number of times G and C appear in the string</span>

<span class="n">Tm_primer</span> <span class="o">=</span> <span class="mf">64.9</span> <span class="o">+</span> <span class="mi">41</span> <span class="o">*</span> <span class="p">(</span><span class="n">GC_primer</span> <span class="o">-</span> <span class="mf">16.4</span><span class="p">)</span><span class="o">/</span><span class="n">N_primer</span>   <span class="c1">#calculate the melting temperature using the formula</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The melting temperature for primer&quot;</span><span class="p">,</span> <span class="n">primer</span><span class="p">,</span> <span class="s2">&quot;is&quot;</span><span class="p">,</span> <span class="n">Tm_primer</span><span class="p">,</span> <span class="s2">&quot;°C .&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The melting temperature for primer GACTGCGTTAGGATTGGC is 50.32222222222223 °C .
</pre></div>
</div>
</div>
</div>
<p><strong>Determine the data type and length for this list with substrate concentrations. Sort the list in ascending order.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">subconc</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">250</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">125</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">60</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">type_subconc</span> <span class="o">=</span> <span class="nb">type</span><span class="p">(</span><span class="n">subconc</span><span class="p">)</span>   <span class="c1">#determine and print the data type</span>

<span class="n">len_subconc</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">subconc</span><span class="p">)</span>   <span class="c1">#determine the length of the list</span>
<span class="n">subconcsorted</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">subconc</span><span class="p">)</span>   <span class="c1">#sort the list in ascending order</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;This variable has data type&quot;</span><span class="p">,</span> <span class="n">type_subconc</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span> 
      <span class="s2">&quot;This list contains&quot;</span><span class="p">,</span> <span class="n">len_subconc</span><span class="p">,</span> <span class="s2">&quot;concentrations .&quot;</span><span class="p">,</span>
      <span class="s2">&quot;After sorting the concentrations in ascending order, this list becomes&quot;</span><span class="p">,</span> <span class="n">subconcsorted</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>This variable has data type &lt;class &#39;list&#39;&gt; . This list contains 11 concentrations . After sorting the concentrations in ascending order, this list becomes [0, 1, 2, 4, 8, 15, 30, 60, 125, 250, 500] .
</pre></div>
</div>
</div>
</div>
<p><strong>Determine the number of “key : value” pairs in the following dictionary of restriction enzymes, with key = name and value = cleavage site. Retrieve the cleavage site for NotI.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">REs</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;EcoRI&#39;</span> <span class="p">:</span> <span class="s1">&#39;GAATTC&#39;</span><span class="p">,</span>
    <span class="s1">&#39;BamHI&#39;</span> <span class="p">:</span> <span class="s1">&#39;GGATCC&#39;</span><span class="p">,</span>
    <span class="s1">&#39;EarI&#39;</span> <span class="p">:</span> <span class="s1">&#39;CTCTTC&#39;</span><span class="p">,</span>
    <span class="s1">&#39;ScaI&#39;</span> <span class="p">:</span> <span class="s1">&#39;AGTACT&#39;</span><span class="p">,</span>
    <span class="s1">&#39;NotI&#39;</span> <span class="p">:</span> <span class="s1">&#39;GCGGCCGC&#39;</span><span class="p">,</span>
    <span class="s1">&#39;TaqI&#39;</span> <span class="p">:</span> <span class="s1">&#39;TCGA&#39;</span><span class="p">,</span>
    <span class="s1">&#39;FokI&#39;</span> <span class="p">:</span> <span class="s1">&#39;GGATG&#39;</span><span class="p">,</span>
    <span class="s1">&#39;HindIII&#39;</span> <span class="p">:</span> <span class="s1">&#39;AAGCTT&#39;</span>
<span class="p">}</span>   <span class="c1">#Create a dictionary with restriction enzyme names (keys) and cleavage sites (values). Both are strings.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">len_REs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">REs</span><span class="p">)</span>   <span class="c1">#determine the length of the dictionary</span>

<span class="n">REs_NotI</span> <span class="o">=</span> <span class="n">REs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;NotI&#39;</span><span class="p">)</span>   <span class="c1">#look up the cleavage site stored for NotI</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;This dictionary contains&quot;</span><span class="p">,</span> <span class="n">len_REs</span><span class="p">,</span> <span class="s2">&quot;pairs .&quot;</span><span class="p">,</span>
      <span class="s2">&quot;The cleavage site stored for NotI is&quot;</span><span class="p">,</span> <span class="n">REs_NotI</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>This dictionary contains 8 pairs . The cleavage site stored for NotI is GCGGCCGC .
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="for-loops-logic-statements-and-functions">
<h1><span style="color:blue">For loops, logic statements, and functions</span><a class="headerlink" href="#for-loops-logic-statements-and-functions" title="Permalink to this heading">#</a></h1>
<hr class="docutils" />
<p><strong>Calculate the <span class="math notranslate nohighlight">\(pK_{b}\)</span> values of aspartate using its <span class="math notranslate nohighlight">\(pK_{a}\)</span> values (2.10, 3.86, 9.82) and save them in a new list.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pKa_asp</span> <span class="o">=</span> <span class="p">[</span><span class="mf">2.10</span><span class="p">,</span> <span class="mf">3.86</span><span class="p">,</span> <span class="mf">9.82</span><span class="p">]</span>   <span class="c1">#create a list with floats</span>

<span class="n">pKb_asp</span> <span class="o">=</span> <span class="p">[]</span>   <span class="c1">#create a new, empty list </span>

<span class="k">for</span> <span class="n">pKa</span> <span class="ow">in</span> <span class="n">pKa_asp</span><span class="p">:</span>   <span class="c1">#select each item from the existing list</span>
    <span class="n">pKb</span> <span class="o">=</span> <span class="mi">14</span> <span class="o">-</span> <span class="n">pKa</span>   <span class="c1">#calculate for each item from the existing list the pKb</span>
    <span class="n">pKb_asp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pKb</span><span class="p">)</span>   <span class="c1">#append the pKb calculated to the newly created list</span>

<span class="nb">print</span><span class="p">(</span><span class="n">pKb_asp</span><span class="p">)</span>   <span class="c1">#print the list that we calculated</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[11.9, 10.14, 4.18]
</pre></div>
</div>
</div>
</div>
<p><strong>Create <em>and test</em> a function that calculates the concentration of a solution using the Beer-Lambert law. Parameters include the molar extinction coefficient, absorbance, and cuvette pathlength.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">beer_lambert</span><span class="p">(</span><span class="n">epsilon</span><span class="p">,</span> <span class="n">absorbance</span><span class="p">,</span> <span class="n">pathlength</span><span class="p">):</span>   <span class="c1">#create the function</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the concentration of a solution, using the Beer-Lambert law.</span>

<span class="sd">    Args:</span>
<span class="sd">        epsilon (float) in L/(mol cm)</span>
<span class="sd">        absorbance (float) in AU</span>
<span class="sd">        pathlength (float) in cm</span>

<span class="sd">    Returns:</span>
<span class="sd">        concentration of the solution (float) in (mol/L)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">concentration</span> <span class="o">=</span> <span class="n">absorbance</span> <span class="o">/</span> <span class="p">(</span><span class="n">epsilon</span> <span class="o">*</span> <span class="n">pathlength</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">concentration</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">beer_lambert</span><span class="p">(</span><span class="mi">21000</span><span class="p">,</span> <span class="mf">0.89</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>4.238095238095238e-05
</pre></div>
</div>
</div>
</div>
<p><strong>Create <em>and test</em> a function that gives the charge of a protein. Parameters include the pH of the solution and the isoelectric point of the protein.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">prot_charge</span><span class="p">(</span><span class="n">pH</span><span class="p">,</span> <span class="n">pI</span><span class="p">):</span>   <span class="c1">#create the function</span>
    <span class="k">if</span> <span class="n">pH</span> <span class="o">&gt;</span> <span class="n">pI</span><span class="p">:</span>   <span class="c1">#if greater then</span>
      <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;pH is above pI, the protein is negatively charged&quot;</span><span class="p">)</span>   <span class="c1">#print the outcome</span>
    <span class="k">elif</span> <span class="n">pH</span> <span class="o">==</span> <span class="n">pI</span><span class="p">:</span>   <span class="c1">#if equal to</span>
      <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;pH and pI are equal, the protein has no net charge&quot;</span><span class="p">)</span>   <span class="c1">#print the outcome</span>
    <span class="k">else</span><span class="p">:</span>   <span class="c1">#if smaller then</span>
      <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;pH is below pI, the protein is positively charged&quot;</span><span class="p">)</span>   <span class="c1">#print the outcome</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">prot_charge</span><span class="p">(</span><span class="mf">5.4</span><span class="p">,</span> <span class="mf">7.5</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>pH is below pI, the protein is positively charged
</pre></div>
</div>
</div>
</div>
<p><strong>Count the number of restriction enzyme sites for EcoRI, BamHI, EarI, ScaI, NotI, TaqI, FokI, and HindIII in the following DNA sequence.</strong></p>
<p><em>Tip: Use the dictionary from earlier!</em></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DNAseq</span> <span class="o">=</span> <span class="s2">&quot;GTAAAACGACGGCCAGTGAATTCGAGCTCGGTACCCGGGGATCCTCTAGAGTCGACCTGCAGGCATGCAAGCTTGGCGTAATCATGGTCATAGCTGTTTCCTG&quot;</span>   <span class="c1">#create a string using double quotation marks</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">REs</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;EcoRI&#39;</span> <span class="p">:</span> <span class="s1">&#39;GAATTC&#39;</span><span class="p">,</span>
    <span class="s1">&#39;BamHI&#39;</span> <span class="p">:</span> <span class="s1">&#39;GGATCC&#39;</span><span class="p">,</span>
    <span class="s1">&#39;EarI&#39;</span> <span class="p">:</span> <span class="s1">&#39;CTCTTC&#39;</span><span class="p">,</span>
    <span class="s1">&#39;ScaI&#39;</span> <span class="p">:</span> <span class="s1">&#39;AGTACT&#39;</span><span class="p">,</span>
    <span class="s1">&#39;NotI&#39;</span> <span class="p">:</span> <span class="s1">&#39;GCGGCCGC&#39;</span><span class="p">,</span>
    <span class="s1">&#39;TaqI&#39;</span> <span class="p">:</span> <span class="s1">&#39;TCGA&#39;</span><span class="p">,</span>
    <span class="s1">&#39;FokI&#39;</span> <span class="p">:</span> <span class="s1">&#39;GGATG&#39;</span><span class="p">,</span>
    <span class="s1">&#39;HindIII&#39;</span> <span class="p">:</span> <span class="s1">&#39;AAGCTT&#39;</span>
<span class="p">}</span>   <span class="c1">#Create a dictionary with restriction enzyme names (keys) and cleavage sites (values). Both are strings.</span>

<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">REs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>   <span class="c1">#select for each item from the existing dictionary key and value</span>
    <span class="n">count</span> <span class="o">=</span> <span class="n">DNAseq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>   <span class="c1">#count the number of times the site appears in the string</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;For&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s2">&quot;, there are&quot;</span><span class="p">,</span> <span class="n">count</span><span class="p">,</span> <span class="s2">&quot;sites .&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>For EcoRI , there are 1 sites .
For BamHI , there are 1 sites .
For EarI , there are 0 sites .
For ScaI , there are 0 sites .
For NotI , there are 0 sites .
For TaqI , there are 2 sites .
For FokI , there are 0 sites .
For HindIII , there are 1 sites .
</pre></div>
</div>
</div>
</div>
<p><strong>Create and test a function that accepts a DNA sequence and outputs the complementary and reverse complementary strand.</strong></p>
<p><em>Tip: To complement a DNA sequence, one needs to replace A with T, T with A, C with G and G with C. The reverse complement is the reverse of the complement sequence (i.e. the first base becomes the last). For example, for the strand 5’-AAGCCGA-3’, the complementary strand is 3’-TTCGGCT-5’ and the reverse complementary strand is 5’-TCGGCTT-3’.</em></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">comp_revcomp</span><span class="p">(</span><span class="n">DNAstring</span><span class="p">):</span>   <span class="c1">#create the function</span>
    
    <span class="n">Comp</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>   <span class="c1">#create a new, empty string</span>
    
    <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">DNAstring</span><span class="p">:</span>   <span class="c1">#select each item from the DNAstring</span>
        <span class="k">if</span> <span class="n">base</span> <span class="o">==</span> <span class="s2">&quot;T&quot;</span><span class="p">:</span>   <span class="c1">#if the base is T</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;A&quot;</span>   <span class="c1">#replace it by A and add to the new string</span>
        <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s2">&quot;A&quot;</span><span class="p">:</span>   <span class="c1">#if the base is T</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;T&quot;</span>   <span class="c1">#replace it by T and add to the new string</span>
        <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s2">&quot;C&quot;</span><span class="p">:</span>   <span class="c1">#if the base is C</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;G&quot;</span>   <span class="c1">#replace it by G and add to the new string</span>
        <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s2">&quot;G&quot;</span><span class="p">:</span>   <span class="c1">#if the base is G</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;C&quot;</span>   <span class="c1">#replace it by C and add to the new string</span>
        <span class="k">else</span><span class="p">:</span>   <span class="c1">#if the base is not T, A, C, or G</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;N&quot;</span>   <span class="c1">#replace it by N and add to the new string</span>
    
    <span class="n">revComp</span> <span class="o">=</span> <span class="n">Comp</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>   <span class="c1">#reverse the new string: start at the end of the string and end at position 0, move with step -1, which is one step backwards</span>
            
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;For 5&#39;-&quot;</span><span class="p">,</span> <span class="n">DNAstring</span><span class="p">,</span> <span class="s2">&quot;-3&#39;, the complement is 3&#39;-&quot;</span><span class="p">,</span> <span class="n">Comp</span><span class="p">,</span> <span class="s2">&quot;-5&#39; .&quot;</span><span class="p">,</span> <span class="s2">&quot;The reverse complement is 5&#39;-&quot;</span><span class="p">,</span> <span class="n">revComp</span><span class="p">,</span> <span class="s2">&quot;-3&#39; .&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">comp_revcomp</span><span class="p">(</span><span class="s2">&quot;AAGCCGA&quot;</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>For 5&#39;- AAGCCGA -3&#39;, the complement is 3&#39;- TTCGGCT -5&#39; . The reverse complement is 5&#39;- TCGGCTT -3&#39; .
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#The following code allows for the bases to be in capital or small letter.</span>
<span class="k">def</span> <span class="nf">comp_revcomp</span><span class="p">(</span><span class="n">DNAstring</span><span class="p">):</span>
    
    <span class="n">Comp</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    
    <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">DNAstring</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">base</span> <span class="ow">in</span> <span class="s2">&quot;Tt&quot;</span><span class="p">:</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;A&quot;</span>
        <span class="k">elif</span> <span class="n">base</span> <span class="ow">in</span> <span class="s2">&quot;Aa&quot;</span><span class="p">:</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;T&quot;</span>
        <span class="k">elif</span> <span class="n">base</span> <span class="ow">in</span> <span class="s2">&quot;Cc&quot;</span><span class="p">:</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;G&quot;</span>
        <span class="k">elif</span> <span class="n">base</span> <span class="ow">in</span> <span class="s2">&quot;Gg&quot;</span><span class="p">:</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;C&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">Comp</span> <span class="o">=</span> <span class="n">Comp</span> <span class="o">+</span> <span class="s2">&quot;N&quot;</span>
    
    <span class="n">revComp</span> <span class="o">=</span> <span class="n">Comp</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;For 5&#39;-&quot;</span><span class="p">,</span> <span class="n">DNAstring</span><span class="p">,</span> <span class="s2">&quot;-3&#39;, the complement is 3&#39;-&quot;</span><span class="p">,</span> <span class="n">Comp</span><span class="p">,</span> <span class="s2">&quot;-5&#39; .&quot;</span><span class="p">,</span> <span class="s2">&quot;The reverse complement is 5&#39;-&quot;</span><span class="p">,</span> <span class="n">revComp</span><span class="p">,</span> <span class="s2">&quot;-3&#39; .&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">comp_revcomp</span><span class="p">(</span><span class="s2">&quot;AaGCcGA&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>For 5&#39;- AaGCcGA -3&#39;, the complement is 3&#39;- TTCGGCT -5&#39; . The reverse complement is 5&#39;- TCGGCTT -3&#39; .
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">comp_revcomp2</span><span class="p">(</span><span class="n">DNAstring</span><span class="p">):</span>   <span class="c1">#create the function</span>
    
    <span class="n">BaseComp</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;A&#39;</span> <span class="p">:</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span>
        <span class="s1">&#39;T&#39;</span> <span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span>
        <span class="s1">&#39;C&#39;</span> <span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span>
        <span class="s1">&#39;G&#39;</span> <span class="p">:</span> <span class="s1">&#39;C&#39;</span>
    <span class="p">}</span>   <span class="c1">#Create a dictionary with complementary base pairs.</span>

    <span class="n">comp</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">BaseComp</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">DNAstring</span><span class="p">)</span>   <span class="c1">#Loop through the bases of the DNAstring and write the complementary base pair from the dictionary. Convert the generated list to a string using the join function, see https://www.programiz.com/python-programming/methods/string/join.</span>
    <span class="n">revcomp</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>   <span class="c1">#reverse the new string: start at the end of the string and end at position 0, move with step -1, which is one step backwards</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;For 5&#39;-&quot;</span><span class="p">,</span> <span class="n">DNAstring</span><span class="p">,</span> <span class="s2">&quot;-3&#39;, the complement is 3&#39;-&quot;</span><span class="p">,</span> <span class="n">comp</span><span class="p">,</span> <span class="s2">&quot;-5&#39; .&quot;</span><span class="p">,</span> <span class="s2">&quot;The reverse complement is 5&#39;-&quot;</span><span class="p">,</span> <span class="n">revcomp</span><span class="p">,</span> <span class="s2">&quot;-3&#39; .&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">comp_revcomp2</span><span class="p">(</span><span class="s2">&quot;AAGCCGA&quot;</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>For 5&#39;- AAGCCGA -3&#39;, the complement is 3&#39;- TTCGGCT -5&#39; . The reverse complement is 5&#39;- TCGGCTT -3&#39; .
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">comp_revcomp3</span><span class="p">(</span><span class="n">DNAstring</span><span class="p">):</span>   <span class="c1">#create the function</span>
    
    <span class="n">mydictcomp</span> <span class="o">=</span> <span class="nb">str</span><span class="o">.</span><span class="n">maketrans</span><span class="p">(</span><span class="s2">&quot;ATCG&quot;</span><span class="p">,</span><span class="s2">&quot;TAGC&quot;</span><span class="p">)</span>   <span class="c1">#use maketrans() to construct the translate table: A (from string 1) becomes T (from string 2), T becomes A ...</span>

    <span class="n">comp</span> <span class="o">=</span> <span class="n">DNAstring</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">mydictcomp</span><span class="p">)</span>   <span class="c1">#use translate() to perform the translations</span>
    <span class="n">revcomp</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>   <span class="c1">#reverse the new string: start at the end of the string and end at position 0, move with step -1, which is one step backwards</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;For 5&#39;-&quot;</span><span class="p">,</span> <span class="n">DNAstring</span><span class="p">,</span> <span class="s2">&quot;-3&#39;, the complement is 3&#39;-&quot;</span><span class="p">,</span> <span class="n">comp</span><span class="p">,</span> <span class="s2">&quot;-5&#39; .&quot;</span><span class="p">,</span> <span class="s2">&quot;The reverse complement is 5&#39;-&quot;</span><span class="p">,</span> <span class="n">revcomp</span><span class="p">,</span> <span class="s2">&quot;-3&#39; .&quot;</span><span class="p">)</span>   <span class="c1">#Print and place all values that we calculated in a readable sentence. The print function can take more than one object. By default, it separates objects with a space.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">comp_revcomp3</span><span class="p">(</span><span class="s2">&quot;AAGCCGA&quot;</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>For 5&#39;- AAGCCGA -3&#39;, the complement is 3&#39;- TTCGGCT -5&#39; . The reverse complement is 5&#39;- TCGGCTT -3&#39; .
</pre></div>
</div>
</div>
</div>
<p><strong>Create and test a function that gives the molecular weight of a peptide from the one-letter amino acid code and the amino acid monoisotopic masses.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">MonoIsoMassAA</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="mf">71.04</span><span class="p">,</span>
                 <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="mf">103.01</span><span class="p">,</span>
                 <span class="s1">&#39;D&#39;</span><span class="p">:</span> <span class="mf">115.03</span><span class="p">,</span>
                 <span class="s1">&#39;E&#39;</span><span class="p">:</span> <span class="mf">129.04</span><span class="p">,</span>
                 <span class="s1">&#39;F&#39;</span><span class="p">:</span> <span class="mf">147.07</span><span class="p">,</span>
                 <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="mf">57.02</span><span class="p">,</span>
                 <span class="s1">&#39;H&#39;</span><span class="p">:</span> <span class="mf">137.06</span><span class="p">,</span>
                 <span class="s1">&#39;I&#39;</span><span class="p">:</span> <span class="mf">113.08</span><span class="p">,</span>
                 <span class="s1">&#39;K&#39;</span><span class="p">:</span> <span class="mf">128.09</span><span class="p">,</span>
                 <span class="s1">&#39;L&#39;</span><span class="p">:</span> <span class="mf">113.08</span><span class="p">,</span>
                 <span class="s1">&#39;M&#39;</span><span class="p">:</span> <span class="mf">131.04</span><span class="p">,</span>
                 <span class="s1">&#39;N&#39;</span><span class="p">:</span> <span class="mf">114.04</span><span class="p">,</span>
                 <span class="s1">&#39;P&#39;</span><span class="p">:</span> <span class="mf">97.05</span><span class="p">,</span>
                 <span class="s1">&#39;Q&#39;</span><span class="p">:</span> <span class="mf">128.06</span><span class="p">,</span>
                 <span class="s1">&#39;R&#39;</span><span class="p">:</span> <span class="mf">156.10</span><span class="p">,</span>
                 <span class="s1">&#39;S&#39;</span><span class="p">:</span> <span class="mf">87.03</span><span class="p">,</span>
                 <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="mf">101.05</span><span class="p">,</span>
                 <span class="s1">&#39;V&#39;</span><span class="p">:</span> <span class="mf">99.07</span><span class="p">,</span>
                 <span class="s1">&#39;W&#39;</span><span class="p">:</span> <span class="mf">186.08</span><span class="p">,</span>
                 <span class="s1">&#39;Y&#39;</span><span class="p">:</span> <span class="mf">163.06</span>
<span class="p">}</span>   <span class="c1">#Create a dictionary with amino acid name (keys) and monoisotopic mass (values).</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">protmass</span><span class="p">(</span><span class="n">ProteinString</span><span class="p">):</span>   <span class="c1">#create the function</span>

    <span class="n">weight</span> <span class="o">=</span> <span class="mi">0</span>   <span class="c1">#start with 0 g/mol</span>

    <span class="k">for</span> <span class="n">code</span><span class="p">,</span> <span class="n">mass</span> <span class="ow">in</span> <span class="n">MonoIsoMassAA</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>   <span class="c1">#select for each item from the existing dictionary key and value</span>
        <span class="n">count</span> <span class="o">=</span> <span class="n">ProteinString</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">code</span><span class="p">)</span>   <span class="c1">#count the number of times the amino acid appears in the ProteinString</span>
        <span class="n">weight</span> <span class="o">=</span> <span class="n">weight</span> <span class="o">+</span> <span class="n">count</span> <span class="o">*</span> <span class="n">mass</span>   <span class="c1">#Multiply the number of times this amino acid appears with its isotopic mass. Add this number to the current weight.</span>
    
    <span class="k">return</span> <span class="n">weight</span> <span class="o">+</span> <span class="mf">18.01</span>   <span class="c1">#return the weight + 18.01, the mass of water</span>
    
<span class="n">protmass</span><span class="p">(</span><span class="s2">&quot;AVATAR&quot;</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>587.35
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">protmass2</span><span class="p">(</span><span class="n">ProteinString</span><span class="p">):</span>   <span class="c1">#create the function</span>
    
    <span class="n">weight</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">MonoIsoMassAA</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ProteinString</span><span class="p">)</span>   <span class="c1">#Loop through the amino acids of the Proteinstring and get their corresponding monoisotopic mass. These are summed.</span>
    
    <span class="k">return</span> <span class="n">weight</span> <span class="o">+</span> <span class="mf">18.01</span>   <span class="c1">#return the weight + 18.01, the mass of water</span>

<span class="n">protmass2</span><span class="p">(</span><span class="s2">&quot;AVATAR&quot;</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>587.35
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">protmass3</span><span class="p">(</span><span class="n">ProteinString</span><span class="p">):</span>   <span class="c1">#create the function</span>
    
    <span class="n">weight</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">MonoIsoMassAA</span><span class="o">.</span><span class="n">get</span><span class="p">,</span> <span class="n">ProteinString</span><span class="p">))</span>   <span class="c1">#The map() function executes a specified function - in this case monoisotopicMassAA.get - for each amino acid of the Proteinstring. It returns the monoisotopic masses from the dictionary, which are summed.</span>
    
    <span class="k">return</span> <span class="n">weight</span> <span class="o">+</span> <span class="mf">18.01</span>   <span class="c1">#return the weight + 18.01, the mass of water</span>

<span class="n">protmass3</span><span class="p">(</span><span class="s2">&quot;AVATAR&quot;</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>587.35
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="drawing-graphs">
<h1><span style="color:blue">Drawing graphs</span><a class="headerlink" href="#drawing-graphs" title="Permalink to this heading">#</a></h1>
<hr class="docutils" />
<p><strong>You measured specific binding data (in cpm), in duplicate, of six concentrations of a radioligand (in µM) to angiotensin receptors on membranes of cells transfected with an angiotensin receptor gene. Plot the duplicate experimental data. Identify (by eye) the trend in the data and the data point(s) that could be considered (an) outlier(s).</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Lconc</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.125</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="mf">4.0</span><span class="p">]</span>
<span class="n">Cpm1</span> <span class="o">=</span> <span class="p">[</span><span class="mi">784</span><span class="p">,</span> <span class="mi">1774</span><span class="p">,</span> <span class="mi">3272</span><span class="p">,</span> <span class="mi">6306</span><span class="p">,</span> <span class="mi">9311</span><span class="p">,</span> <span class="mi">12168</span><span class="p">]</span>
<span class="n">Cpm2</span> <span class="o">=</span> <span class="p">[</span><span class="mi">792</span><span class="p">,</span> <span class="mi">1645</span><span class="p">,</span> <span class="mi">3169</span><span class="p">,</span> <span class="mi">3349</span><span class="p">,</span> <span class="mi">8756</span><span class="p">,</span> <span class="mi">11730</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the library needed (i.e. matplotlib) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1">#create the graph</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>   <span class="c1">#create a figure object, the size is provided in inches (width, height)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">Lconc</span><span class="p">,</span> <span class="n">Cpm1</span><span class="p">,</span>   <span class="c1">#plot a set of (x (= ligand concentrations in µM),y (= specific binding in cpm)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;g&#39;</span><span class="p">,</span>   <span class="c1">#use a green color</span>
         <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>   <span class="c1">#use a circle as marker</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Dataset 1&quot;</span><span class="p">)</span>   <span class="c1">#use a legend label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">Lconc</span><span class="p">,</span> <span class="n">Cpm2</span><span class="p">,</span>   <span class="c1">#plot a set of (x (= ligand concentrations in µM),y (= specific binding in cpm)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;b&#39;</span><span class="p">,</span>   <span class="c1">#use a blue color</span>
         <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span>   <span class="c1">#use a square as marker</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Dataset 2&quot;</span><span class="p">)</span>   <span class="c1">#use a legend label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Binding of radioligand to angiotensin receptors&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>   <span class="c1">#define the graph title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;$[Ligand]$, µM&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the X-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Specific binding, cpm&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the Y-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>   <span class="c1">#include a legend</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>   <span class="c1">#show the plot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ab9ca7354b88f81b846a66bcda3031a7c823088b061609371f607bcb86e75e92.png" src="../_images/ab9ca7354b88f81b846a66bcda3031a7c823088b061609371f607bcb86e75e92.png" />
</div>
</div>
<p>→ We can see a hyperbolic trend. We have not reached saturation yet: there is no horizontal asymptote for increasing <span class="math notranslate nohighlight">\([Ligand]\)</span>. It looks like there is an outlier for <span class="math notranslate nohighlight">\([Ligand]\)</span> = 1 µM. Statistical tests or weighted fits are needed <em>(more about this later)</em>!</p>
<p><strong>Use the Michaelis-Menten equation. Calculate and plot the initial velocities of an enzyme-catalysed reaction to investigate what happens when <span class="math notranslate nohighlight">\(K_{m}\)</span> is small / large.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the libraries needed (i.e. numpy and matplotlib) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1">#create the function</span>
<span class="k">def</span> <span class="nf">MichaelisMenten</span><span class="p">(</span><span class="n">Sconc</span><span class="p">,</span> <span class="n">Vmax</span><span class="p">,</span> <span class="n">Km</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the initial velocity of an enzyme-catalysed reaction with the Michaelis-Menten equation,</span>
<span class="sd">    using the initial substrate concentration, Vmax, and Km.</span>
<span class="sd">    Args:</span>
<span class="sd">        Sconc (float), the initial substrate concentration in mM or µM or ...</span>
<span class="sd">        Vmax (float), the maximum velocity in mM/min or µM/min or mM/s or µM/s or ...</span>
<span class="sd">        Km (float), the Michaelis-Menten constant in mM or µM or ...</span>
<span class="sd">    Returns:</span>
<span class="sd">        The initial velocity, v0 of an enzyme-catalysed reaction (float) in mM/min or µM/min or mM/s or µM/s or ...</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">v0</span> <span class="o">=</span> <span class="p">(</span><span class="n">Vmax</span> <span class="o">*</span> <span class="n">Sconc</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">Km</span> <span class="o">+</span> <span class="n">Sconc</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">v0</span>

<span class="n">MichaelisMenten</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>30.0
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#create substrate concentrations to use as x-values for our function</span>
<span class="n">substratetest</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">41</span><span class="p">)</span>   <span class="c1">#return 41 evenly spaced numbers between 0 and 20</span>

<span class="c1">#create the graph</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>   <span class="c1">#create a figure object, the size is provided in inches (width, height)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">substratetest</span><span class="p">,</span> <span class="n">MichaelisMenten</span><span class="p">(</span><span class="n">substratetest</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>   <span class="c1">#plot a set of (x (= initial substrate concentrations),y (= the calculated initial velocities using the Vmax and Km provided)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;g&#39;</span><span class="p">,</span>   <span class="c1">#use a green color</span>
         <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span>   <span class="c1">#use a solid line</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;$V_</span><span class="si">{max}</span><span class="s2">$ = 40 µM/min, $K_m$ = 5 µM&quot;</span><span class="p">)</span>   <span class="c1">#use a legend label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">substratetest</span><span class="p">,</span> <span class="n">MichaelisMenten</span><span class="p">(</span><span class="n">substratetest</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span>   <span class="c1">#plot a set of (x (= initial substrate concentrations),y (= the calculated initial velocities using the Vmax and Km provided)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;b&#39;</span><span class="p">,</span>   <span class="c1">#use a blue color</span>
         <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>   <span class="c1">#use a dashed line</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;$V_</span><span class="si">{max}</span><span class="s2">$ = 40 µM/min, $K_m$ = 0.5 µM&quot;</span><span class="p">)</span>   <span class="c1">#use a legend label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Michaelis-Menten curve: the influence of $K_m$&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>   <span class="c1">#define the graph title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;$[Substrate]_0$, µM&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the X-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;$v_0$, µM/min&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the Y-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>   <span class="c1">#include a legend</span>

<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;figure_enzyme.png&#39;</span><span class="p">)</span>   <span class="c1">#save the plot as a figure</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/045ff7fb6373d03d336e0df6d138803a0fd956274701737627d52168b20dbddb.png" src="../_images/045ff7fb6373d03d336e0df6d138803a0fd956274701737627d52168b20dbddb.png" />
</div>
</div>
<p>→ We can see a hyperbolic trend. We can see that a small <span class="math notranslate nohighlight">\(K_{m}\)</span> indicates that the enzyme requires only a small amount of substrate to become saturated: the maximum velocity is reached at relatively low substrate concentrations. A large <span class="math notranslate nohighlight">\(K_{m}\)</span> indicates the need for higher substrate concentrations to achieve maximum reaction velocity.</p>
<p><strong>You measured the number of days to flowering for a wild type and six mutants. Plot the findings.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Plant</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Wild type&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant 1&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant 2&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant 3&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant 4&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant 5&quot;</span><span class="p">,</span> <span class="s2">&quot;Mutant 6&quot;</span><span class="p">]</span>
<span class="n">DaysToFlower</span> <span class="o">=</span> <span class="p">[</span><span class="mi">25</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">80</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">23</span><span class="p">,</span> <span class="mi">60</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the library needed (i.e. matplotlib) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1">#create the graph</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>   <span class="c1">#create a figure object, the size is provided in inches (width, height)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">Plant</span><span class="p">,</span> <span class="n">DaysToFlower</span><span class="p">,</span>   <span class="c1">#plot a set of (x (= wild type or mutants),y (= the measured variable)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;g&#39;</span><span class="p">)</span>   <span class="c1">#use a green color</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">labelrotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>   <span class="c1">#rotate the names for the x-coordinates 90 degrees</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Number of days to flower for wild type and mutant plants&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>   <span class="c1">#define the graph title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of days to flower&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the Y-axis label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>   <span class="c1">#show the plot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1de483aff1f4cb0c41c6afab9879bfe1a49726329a9f9460a15ce6bbbc16f1f0.png" src="../_images/1de483aff1f4cb0c41c6afab9879bfe1a49726329a9f9460a15ce6bbbc16f1f0.png" />
</div>
</div>
<p>→ We can see that some mutant take more time to flower than others.</p>
<p><strong>You measured initial velocities (in µM/s) of an enzyme-catalyzed reaction using eight initial substrate concentrations (in µM) in the absence and presence of 15 µM inhibitor. Plot the experimental data. Identify (by eye) the type of inhibitor.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Sconc</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="mi">32</span><span class="p">,</span> <span class="mi">64</span><span class="p">,</span> <span class="mi">128</span><span class="p">]</span>
<span class="n">NoI</span> <span class="o">=</span> <span class="p">[</span><span class="mi">185</span><span class="p">,</span> <span class="mi">227</span><span class="p">,</span> <span class="mi">327</span><span class="p">,</span> <span class="mi">555</span><span class="p">,</span> <span class="mi">614</span><span class="p">,</span> <span class="mi">757</span><span class="p">,</span> <span class="mi">877</span><span class="p">,</span> <span class="mi">897</span><span class="p">]</span>
<span class="n">WithI</span> <span class="o">=</span> <span class="p">[</span><span class="mi">15</span><span class="p">,</span> <span class="mi">48</span><span class="p">,</span> <span class="mi">155</span><span class="p">,</span> <span class="mi">180</span><span class="p">,</span> <span class="mi">300</span><span class="p">,</span> <span class="mi">404</span><span class="p">,</span> <span class="mi">624</span><span class="p">,</span> <span class="mi">830</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the library needed (i.e. matplotlib) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1">#create the graph</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>   <span class="c1">#create a figure object, the size is provided in inches (width, height)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">Sconc</span><span class="p">,</span> <span class="n">NoI</span><span class="p">,</span>   <span class="c1">#plot a set of (x (= substrate concentrations in µM),y (= initial velocities in µM/s)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;g&#39;</span><span class="p">,</span>   <span class="c1">#use a green color</span>
         <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>   <span class="c1">#use a circle as marker</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Without inhibitor&quot;</span><span class="p">)</span>   <span class="c1">#use a legend label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">Sconc</span><span class="p">,</span> <span class="n">WithI</span><span class="p">,</span>   <span class="c1">#plot a set of (x (= substrate concentrations in µM),y (= initial velocities in µM/s)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;b&#39;</span><span class="p">,</span>   <span class="c1">#use a blue color</span>
         <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span>   <span class="c1">#use a square as marker</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;With inhibitor&quot;</span><span class="p">)</span>   <span class="c1">#use a legend label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Initial velocities in the absence and presence of inhibitor&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>   <span class="c1">#define the graph title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;$[Substrate]$, µM&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the X-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;$v_0$, µM/s&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the Y-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>   <span class="c1">#include a legend</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>   <span class="c1">#show the plot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/a643810fec8c160413887d4501be6d64eccf8ec6d0e4a1b5853ca6d2645118b7.png" src="../_images/a643810fec8c160413887d4501be6d64eccf8ec6d0e4a1b5853ca6d2645118b7.png" />
</div>
</div>
<p>→ <span class="math notranslate nohighlight">\(V_{max}\)</span>, the horizontal asymptote at high <span class="math notranslate nohighlight">\([Substrate]\)</span> looks the same. <span class="math notranslate nohighlight">\(K_{m}\)</span>, the concentration of substrate at which the enzyme achieves half Vmax, looks larger in the presence of the inhibitor. This behaviour is characteristic of a competitive inhibitor.</p>
<p><strong>Use the following equation: fraction of bound receptor = <span class="math notranslate nohighlight">\([RL]\)</span> / (<span class="math notranslate nohighlight">\([RL]\)</span> + <span class="math notranslate nohighlight">\([R]\)</span>) = <span class="math notranslate nohighlight">\([Ligand]\)</span> / (<span class="math notranslate nohighlight">\([Ligand]\)</span> + <span class="math notranslate nohighlight">\(K_{d}\)</span>). Calculate and plot the fraction of bound receptor to investigate what happens when <span class="math notranslate nohighlight">\(K_{d}\)</span> is small / large.</strong></p>
<p>Tips:</p>
<ul class="simple">
<li><p>Define a function that calculates the fraction of bound receptor using the equation provided.</p></li>
<li><p>Create an array using a fixed number of points, <em>e.g.</em> 1000, between 0 and 100 mM. This will serve as the x-values.</p></li>
<li><p>Plot the function for two different scenarios, <em>e.g.</em> <span class="math notranslate nohighlight">\(K_{d}\)</span> = 1 mM and <span class="math notranslate nohighlight">\(K_{d}\)</span> = 10 mM. Use a different color for each curve.</p></li>
<li><p>Make sure to include a title, X-and Y-axes labels, and a legend.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the libraries needed (i.e. numpy and matplotlib) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1">#create the function</span>
<span class="k">def</span> <span class="nf">FractionBoundReceptor</span><span class="p">(</span><span class="n">Lconc</span><span class="p">,</span> <span class="n">Kd</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the fraction of bound receptor using the initial ligand concentration and Kd.</span>
<span class="sd">    Args:</span>
<span class="sd">        Lconc (float), the initial ligand concentration in mM or µM or ...</span>
<span class="sd">        Kd (float), the dissociation constant in mM or µM or ...</span>
<span class="sd">    Returns:</span>
<span class="sd">        The fraction of bound receptor in %</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">FractionBound</span> <span class="o">=</span> <span class="n">Lconc</span> <span class="o">/</span> <span class="p">(</span><span class="n">Lconc</span> <span class="o">+</span> <span class="n">Kd</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">FractionBound</span>

<span class="n">FractionBoundReceptor</span><span class="p">(</span><span class="mi">18</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">)</span>   <span class="c1">#test the function with random but reasonable parameters</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.9230769230769231
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#create substrate concentrations to use as x-values for our function</span>
<span class="n">ligandtest</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>   <span class="c1">#return 1000 evenly spaced numbers between 0 and 100</span>

<span class="c1">#create the graph</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>   <span class="c1">#create a figure object, the size is provided in inches (width, height)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ligandtest</span><span class="p">,</span> <span class="n">FractionBoundReceptor</span><span class="p">(</span><span class="n">ligandtest</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>   <span class="c1">#plot a set of (x (= initial ligand concentrations),y (= the calculated fraction of bound receptor using the Kd provided)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;g&#39;</span><span class="p">,</span>   <span class="c1">#use a green color</span>
         <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span>   <span class="c1">#use a solid line</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;$K_d$ = 1 mM&quot;</span><span class="p">)</span>   <span class="c1">#use a legend label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ligandtest</span><span class="p">,</span> <span class="n">FractionBoundReceptor</span><span class="p">(</span><span class="n">ligandtest</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span>   <span class="c1">#plot a set of (x (= initial ligand concentrations),y (= the calculated fraction of bound receptor using the Kd provided)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;b&#39;</span><span class="p">,</span>   <span class="c1">#use a blue color</span>
         <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>   <span class="c1">#use a dashed line</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;$K_d$ = 10 mM&quot;</span><span class="p">)</span>   <span class="c1">#use a legend label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Fraction of bound receptor: the influence of $K_d$&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>   <span class="c1">#define the graph title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;$[Ligand]_0$, µM&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the X-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Fraction of bound receptor, %&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the Y-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>   <span class="c1">#include a legend</span>

<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;figure_binding.png&#39;</span><span class="p">)</span>   <span class="c1">#save the plot as a figure</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/295bfc667feafedea7d9970d52e80b65916d52b3a584c3c1d80f93c1ffef530c.png" src="../_images/295bfc667feafedea7d9970d52e80b65916d52b3a584c3c1d80f93c1ffef530c.png" />
</div>
</div>
<p>→ When <span class="math notranslate nohighlight">\([Ligand]_{0}\)</span> = 0, no receptor has ligand bound. When <span class="math notranslate nohighlight">\([Ligand]_{0}\)</span> = <span class="math notranslate nohighlight">\(K_{d}\)</span>, the fractional occupancy is 0.50. When <span class="math notranslate nohighlight">\([Ligand]_{0}\)</span> =9 <span class="math notranslate nohighlight">\(K_{d}\)</span>, the fractional occupancy is 0.90.</p>
<p>We can see that a small <span class="math notranslate nohighlight">\(K_{d}\)</span> indicates that the receptor requires only a small amount of ligand to become saturated. A large Kd indicates the need for higher ligand concentrations to achieve saturation.</p>
<p><strong>You measured specific activities (in U/mg) for wild type and mutant enzyme. Plot the experimental data. Which mutant is the most / least active?</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Protein</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Wild type&quot;</span><span class="p">,</span> <span class="s2">&quot;Y46A&quot;</span><span class="p">,</span> <span class="s2">&quot;W50A&quot;</span><span class="p">,</span> <span class="s2">&quot;R55A&quot;</span><span class="p">,</span> <span class="s2">&quot;W171A&quot;</span><span class="p">,</span> <span class="s2">&quot;D85A&quot;</span><span class="p">]</span>
<span class="n">SpecificActivity</span> <span class="o">=</span> <span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">83</span><span class="p">,</span> <span class="mi">45</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">62</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the library needed (i.e. matplotlib) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1">#create the graph</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>   <span class="c1">#create a figure object, the size is provided in inches (width, height)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">Protein</span><span class="p">,</span> <span class="n">SpecificActivity</span><span class="p">,</span>   <span class="c1">#plot a set of (x (= wild type or mutants),y (= the measured variable)) data</span>
         <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="s1">&#39;darkblue&#39;</span><span class="p">,</span> <span class="s1">&#39;turquoise&#39;</span><span class="p">,</span> <span class="s1">&#39;mediumvioletred&#39;</span><span class="p">,</span> <span class="s1">&#39;orange&#39;</span><span class="p">,</span> <span class="s1">&#39;violet&#39;</span><span class="p">])</span>   <span class="c1">#use a green color</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">labelrotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>   <span class="c1">#rotate the names for the x-coordinates 90 degrees</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Enzyme activity for wild type and mutant enzymes&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>   <span class="c1">#define the graph title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Specific activity (U/mg)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>   <span class="c1">#define the Y-axis label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>   <span class="c1">#show the plot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/fcdc2d8672ff5dc86f40a425b23240cd17c7a7f1615394f31024c9221e87004a.png" src="../_images/fcdc2d8672ff5dc86f40a425b23240cd17c7a7f1615394f31024c9221e87004a.png" />
</div>
</div>
<p>→ We can see that W50A is the most active and Y46A the least active mutant.</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="pandas-for-working-with-tabular-data">
<h1><span style="color:blue">Pandas For Working With Tabular Data</span><a class="headerlink" href="#pandas-for-working-with-tabular-data" title="Permalink to this heading">#</a></h1>
<hr class="docutils" />
<p><strong>You measured the BSA standards for a colorimetric assay in duplicate. You collected your data in an Excel file called “StandardCurve.xlsx”. Read in this data file with pandas and plot the data with matplotlib.</strong></p>
<p>Download the Excel file <a class="reference external" href="https://github.com/renee-85/PythonForBiochemists/tree/master/data">here</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the library needed (i.e. pandas) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">dfStandardCurve</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span> <span class="p">(</span><span class="s1">&#39;../data/StandardCurve.xlsx&#39;</span><span class="p">,</span>   <span class="c1">#create a pandas DataFrame from the filename with file path and extension shown</span>
                    <span class="n">sheet_name</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>   <span class="c1">#use the first sheet (no need to specifically include this as we use the default setting)</span>
                    <span class="n">skiprows</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>   <span class="c1">#skip no rows (no need to specifically include this as we use the default setting)</span>
                    <span class="n">usecols</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>   <span class="c1">#import all columns (no need to specifically include this as we use the default setting)</span>
                    <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>   <span class="c1">#use the column names from the now first row as column labels (no need to specifically include this as we use the default setting)</span>
                    <span class="n">skipfooter</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>   <span class="c1">#skip no footers (no need to specifically include this as we use the default setting)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">dfStandardCurve</span><span class="p">)</span>   <span class="c1">#print the DataFrame created</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>   [BSA] (mg/ml)  A562nm (AU) - 1  A562nm (AU) - 2
0          2.000            1.405            1.240
1          1.500            1.125            0.974
2          1.000            0.844            0.686
3          0.750            0.705            0.560
4          0.500            0.539            0.436
5          0.250            0.311            0.278
6          0.125            0.209            0.188
7          0.000            0.100            0.095
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the library needed (i.e. pandas) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1">#Plot the data</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>   <span class="c1">#start a figure object</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">dfStandardCurve</span><span class="p">[</span><span class="s1">&#39;[BSA] (mg/ml)&#39;</span><span class="p">],</span> <span class="n">dfStandardCurve</span><span class="p">[</span><span class="s1">&#39;A562nm (AU) - 1&#39;</span><span class="p">],</span>   <span class="c1">#plot a set of x (= the BSA concentrations) ,y (=the measured absorbances) data</span>
         <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>   <span class="c1">#use a round, green marker with size 8</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">dfStandardCurve</span><span class="p">[</span><span class="s1">&#39;[BSA] (mg/ml)&#39;</span><span class="p">],</span> <span class="n">dfStandardCurve</span><span class="p">[</span><span class="s1">&#39;A562nm (AU) - 2&#39;</span><span class="p">],</span>   <span class="c1">#plot a set of x (= the same BSA concentrations) ,y (=the measured absorbances) data, a duplicate experiment</span>
         <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>   <span class="c1">#use a square, blue marker with size 8</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Standard Curve&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>   <span class="c1">#define the graph title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;$[BSA]$ (mg ml$^{-1}$)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>   <span class="c1">#define the X-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;$A_</span><span class="si">{562nm}</span><span class="s1">$ (AU)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>   <span class="c1">#define the Y-axis label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>   <span class="c1">#show the figure object</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b20d7fbef2e2a2039a4f48e60b10fc0ab0dd796e11c6c03a3ea5c1692ee7d157.png" src="../_images/b20d7fbef2e2a2039a4f48e60b10fc0ab0dd796e11c6c03a3ea5c1692ee7d157.png" />
</div>
</div>
<p><strong>You measured absorbance versus time for an enzyme reaction. You collected your data in an Excel file called “EnzymeKinetics.xlsx”. Read in this data file with pandas and plot the data with matplotlib.</strong></p>
<p>Download the Excel file <a class="reference external" href="https://github.com/renee-85/PythonForBiochemists/tree/master/data">here</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the library needed (i.e. pandas) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">dfEnzymeKinetics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span> <span class="p">(</span><span class="s1">&#39;../data/EnzymeKinetics.xlsx&#39;</span><span class="p">,</span>   <span class="c1">#create a pandas DataFrame from the filename with file path and extension shown</span>
                    <span class="n">sheet_name</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>   <span class="c1">#use the first sheet (no need to specifically include this as we use the default setting)</span>
                    <span class="n">skiprows</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>   <span class="c1">#skip no rows (no need to specifically include this as we use the default setting)</span>
                    <span class="n">usecols</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>   <span class="c1">#import all columns (no need to specifically include this as we use the default setting)</span>
                    <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>   <span class="c1">#use the column names from the now first row as column labels (no need to specifically include this as we use the default setting)</span>
                    <span class="n">skipfooter</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>   <span class="c1">#skip no footers (no need to specifically include this as we use the default setting)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">dfEnzymeKinetics</span><span class="p">)</span>   <span class="c1">#print the DataFrame created</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    Time (s)  A450nm (AU)
0          0        0.052
1          5        0.052
2         10        0.052
3         15        0.052
4         20        0.052
5         25        0.052
6         30        0.053
7         35        0.053
8         40        0.054
9         45        0.055
10        50        0.056
11        55        0.057
12        60        0.058
13        65        0.059
14        70        0.060
15        75        0.061
16        80        0.063
17        85        0.064
18        90        0.065
19        95        0.066
20       100        0.067
21       105        0.068
22       110        0.069
23       115        0.069
24       120        0.070
25       125        0.071
26       130        0.073
27       135        0.074
28       140        0.075
29       145        0.076
30       150        0.077
31       155        0.079
32       160        0.080
33       165        0.081
34       170        0.082
35       175        0.083
36       180        0.084
37       185        0.085
38       190        0.086
39       195        0.087
40       200        0.088
41       205        0.089
42       210        0.091
43       216        0.092
44       221        0.093
45       226        0.094
46       231        0.095
47       236        0.096
48       241        0.098
49       246        0.099
50       251        0.100
51       256        0.102
52       261        0.103
53       266        0.105
54       271        0.106
55       276        0.107
56       281        0.108
57       286        0.109
58       291        0.111
59       296        0.111
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import the library needed (i.e. pandas) and introduce convenient naming</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1">#Plot the data</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>   <span class="c1">#start a figure object</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">dfEnzymeKinetics</span><span class="p">[</span><span class="s1">&#39;Time (s)&#39;</span><span class="p">],</span> <span class="n">dfEnzymeKinetics</span><span class="p">[</span><span class="s1">&#39;A450nm (AU)&#39;</span><span class="p">],</span>   <span class="c1">#plot a set of x (= time points) ,y (=the measured absorbances) data</span>
         <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>   <span class="c1">#use a round, black marker with size 8</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Progress curve&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>   <span class="c1">#define the graph title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;$Time$ (s)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>   <span class="c1">#define the X-axis label</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;$A_</span><span class="si">{450nm}</span><span class="s1">$ (AU)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>   <span class="c1">#define the Y-axis label</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>   <span class="c1">#show the figure object</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/10b62583e9eef63995a21c56daec79102d0222fdddcdb1d18dc7605d8dd813d5.png" src="../_images/10b62583e9eef63995a21c56daec79102d0222fdddcdb1d18dc7605d8dd813d5.png" />
</div>
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./chapters"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#"><span style="color:blue">Variables and arithmetic operators</span></a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-types-index-and-slice"><span style="color:blue">Data types, index, and slice</span></a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#for-loops-logic-statements-and-functions"><span style="color:blue">For loops, logic statements, and functions</span></a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#drawing-graphs"><span style="color:blue">Drawing graphs</span></a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#pandas-for-working-with-tabular-data"><span style="color:blue">Pandas For Working With Tabular Data</span></a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Charmian Dawson, Toryn Poolman, and Renee Vancraenenbroeck; UCL
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>